File: notifications-service/swagger_server/vault_loader.py
import os
import requests
import logging

# Configura o logger para aparecer nos logs do Docker
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("VaultLoader")

def load_vault_secrets():
    vault_addr = os.getenv("VAULT_ADDR", "http://vault:8200")
    vault_token = os.getenv("VAULT_TOKEN", "dev-root")
    # Path para KV V2
    vault_path = os.getenv("VAULT_SECRET_PATH", "secret/data/compositor")

    logger.info(f"Connecting to Vault at {vault_addr}...")

    try:
        url = f"{vault_addr}/v1/{vault_path}"
        headers = {"X-Vault-Token": vault_token}
        
        response = requests.get(url, headers=headers, timeout=5)
        
        if response.status_code == 200:
            data = response.json()
            # No motor KV V2, os segredos estão em data['data']
            secrets = data.get("data", {}).get("data", {})
            
            count = 0
            for key, value in secrets.items():
                # Injeta no ambiente se não existir (ou estiver vazio)
                if not os.getenv(key):
                    os.environ[key] = str(value)
                    count += 1
            
            logger.info(f"Successfully loaded {count} secrets from Vault.")
        else:
            logger.error(f"Failed to load secrets. Status: {response.status_code}")
            
    except Exception as e:
        logger.error(f"Error connecting to Vault: {e}")

# Executa se o ficheiro for chamado diretamente
if __name__ == "__main__":
    load_vault_secrets()

File: oam-service/src/vault_loader.rs
use std::env;
use std::collections::HashMap;

pub fn load_vault_secrets() {
    let vault_addr = env::var("VAULT_ADDR").unwrap_or_else(|_| "http://vault:8200".to_string());
    let vault_token = env::var("VAULT_TOKEN").unwrap_or_else(|_| "dev-root".to_string());
    let vault_path = env::var("VAULT_SECRET_PATH").unwrap_or_else(|_| "secret/data/compositor".to_string());

    println!("[Vault] Attempting to load secrets from {}...", vault_addr);

    let client = reqwest::blocking::Client::new();
    let url = format!("{}/v1/{}", vault_addr, vault_path);

    let res = client
        .get(url)
        .header("X-Vault-Token", vault_token)
        .send();

    match res {
        Ok(response) => {
            if response.status().is_success() {
                let json: serde_json::Value = response.json().unwrap_or_default();
                
                // No Vault KV V2, os dados estão em ["data"]["data"]
                if let Some(secrets) = json["data"]["data"].as_object() {
                    let mut count = 0;
                    for (key, value) in secrets {
                        // Só injeta se a variável não estiver já definida no sistema
                        if env::var(key).is_err() {
                            let val_str = value.as_str().unwrap_or("");
                            env::set_var(key, val_str);
                            count += 1;
                        }
                    }
                    println!("[Vault] Successfully loaded {} secrets.", count);
                }
            } else {
                println!("[Vault] Error: Failed to load secrets. Status: {}", response.status());
            }
        }
        Err(e) => println!("[Vault] Connection error: {}", e),
    }
}